src/main/java/org/apache/sysml/hops/OptimizerUtils.java

```diff
@@ -157,6 +157,18 @@
 	public static boolean ALLOW_INTER_PROCEDURAL_ANALYSIS = true;
 
 	/**
+	 * Enables an additional "second chance" pass of static rewrites + IPA after the initial pass of
+	 * IPA.  Without this, there are many situations in which sizes will remain unknown even after
+	 * recompilation, thus leading to distributed ops.  With the second chance enabled, sizes in
+	 * these situations can be determined.  For example, the alternation of constant folding
+	 * (static rewrite) and scalar replacement (IPA) can allow for size propagation without dynamic
+	 * rewrites or recompilation due to replacement of scalars with literals during IPA, which
+	 * enables constant folding of sub-DAGs of literals during static rewrites, which in turn allows
+	 * for scalar propagation during IPA.
+	 */
+	public static boolean ALLOW_IPA_SECOND_CHANCE = true;
```

Comment:
I would recommend >=2 (which is what your code currently implicitly does because it is conditioned on inter procedural analysis) but it would be good to make it explicit